File: erubit.platform/src/main/java/a/erubit/platform/android/ProgressFragment.java
package a.erubit.platform.android;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.StaggeredGridLayoutManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import a.erubit.platform.R;
import a.erubit.platform.course.Course;
import a.erubit.platform.course.CourseManager;
import a.erubit.platform.course.Lesson;
import a.erubit.platform.course.SetLesson;
import a.erubit.platform.course.WelcomeLesson;

public class ProgressFragment extends Fragment {

    private LessonsFragment.OnLessonInteractionListener mListener;

    public ProgressFragment() {
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_progress, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        RecyclerView listView = view.findViewById(android.R.id.list);
        StaggeredGridLayoutManager layoutManager = new StaggeredGridLayoutManager(1, StaggeredGridLayoutManager.VERTICAL);
        listView.setItemAnimator(new DefaultItemAnimator());
        listView.setLayoutManager(layoutManager);

        final Course course = CourseManager.i().getCourse(getArguments().getString("id"));
        ((TextView)view.findViewById(R.id.courseName)).setText(course.name);
        listView.setAdapter(new ProgressListAdapter(course));
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        if (context instanceof LessonsFragment.OnLessonInteractionListener)
            mListener = (LessonsFragment.OnLessonInteractionListener) context;
    }


    private class ProgressListAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
        private final List<ListItem> mList;

        ProgressListAdapter(final Course course) {
            mList = new ArrayList<>(10);

            for (Lesson lesson: course.getLessons()) {
                mList.add(new HeaderItem(lesson));
                if (lesson instanceof WelcomeLesson) {
                    mList.add(new ContentItem(lesson.getProgress().getExplanation(), ""));
                }
                if (lesson instanceof SetLesson) { // also PhraseLesson and CharacterLesson here
                    ((SetLesson)lesson).loadHeavyContent();
                    for (SetLesson.Item item: ((SetLesson)lesson).mSet) {
                        mList.add(new ContentItem(item.character, getKnowledgeText(item.knowledgeLevel)));
                    }
                }
            }
        }

        private String getKnowledgeText(int knowledgeLevel) {
            if (knowledgeLevel > 0 && knowledgeLevel < SetLesson.RANK_FAMILIAR)
                return App.getContext().getString(R.string.studying);
            else if (knowledgeLevel >= SetLesson.RANK_FAMILIAR && knowledgeLevel < SetLesson.RANK_LEARNED)
                return App.getContext().getString(R.string.familiar);
            else if (knowledgeLevel >= SetLesson.RANK_LEARNED && knowledgeLevel < SetLesson.RANK_LEARNED_WELL)
                return App.getContext().getString(R.string.learned);
            else if (knowledgeLevel >= SetLesson.RANK_LEARNED_WELL)
                return App.getContext().getString(R.string.learned_well);
            else
                return App.getContext().getString(R.string.unknown);
        }

        @Override
        public int getItemViewType(int position) {
            return mList.get(position).getType();
        }

        @Override
        public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            LayoutInflater inflater = LayoutInflater.from(parent.getContext());

            switch (viewType) {
                case ListItem.HEADER:
                    View view = inflater.inflate(R.layout.item_progress_header, parent, false);
                    return new HeaderViewHolder(view);
                case ListItem.CONTENT:
                    view = inflater.inflate(R.layout.item_progress_content, parent, false);
                    return new ContentViewHolder(view);
                default:
                    return null;
            }
        }

        @Override
        public void onBindViewHolder(final RecyclerView.ViewHolder holder, int position) {
            if (holder instanceof HeaderViewHolder) {
                ((HeaderViewHolder) holder).mLesson = ((HeaderItem) mList.get(position)).mLesson;
                String text = ((HeaderItem) mList.get(position)).mName;
                ((HeaderViewHolder) holder).textTitle.setText(text);
            }
            if (holder instanceof ContentViewHolder) {
                ContentItem item = ((ContentItem) mList.get(position));
                ((ContentViewHolder) holder).textTitle.setText(item.mText);
                ((ContentViewHolder) holder).textKnowledge.setText(item.mKnowledge);
            }
        }

        @Override
        public int getItemCount() {
            return mList.size();
        }

        private class HeaderViewHolder extends RecyclerView.ViewHolder {
            Lesson mLesson;
            private final TextView textTitle;
            private final Button btnPractice;

            HeaderViewHolder(View itemView) {
                super(itemView);
                
                textTitle = itemView.findViewById(android.R.id.text1);
                btnPractice = itemView.findViewById(android.R.id.button1);
                btnPractice.setOnClickListener(view -> mListener.onLessonInteraction(mLesson, LessonsFragment.LessonInteractionAction.PRACTICE));
            }
        }

        private class ContentViewHolder extends RecyclerView.ViewHolder {
            private final TextView textTitle;
            private final TextView textKnowledge;

            ContentViewHolder(View itemView) {
                super(itemView);

                textTitle = itemView.findViewById(android.R.id.text1);
                textKnowledge = itemView.findViewById(android.R.id.text2);
            }
        }

        abstract class ListItem {
            static final int HEADER = 0;
            static final int CONTENT = 1;

            abstract public int getType();
        }

        private class HeaderItem extends ListItem {
            private final Lesson mLesson;
            private final String mName;

            HeaderItem(Lesson lesson) {
                this.mLesson = lesson;
                this.mName = lesson.name;
            }

            @Override
            public int getType() {
                return HEADER;
            }
        }

        private class ContentItem extends ListItem {
            private final String mText;
            private final String mKnowledge;

            ContentItem(String text, String knowledge) {
                mText = text;
                mKnowledge = knowledge;
            }

            @Override
            public int getType() {
                return CONTENT;
            }
        }
    }
}

File: erubit.platform/src/main/java/a/erubit/platform/android/ProgressFragment.kt
package a.erubit.platform.android

import a.erubit.platform.R
import a.erubit.platform.android.LessonsFragment.OnLessonInteractionListener
import a.erubit.platform.course.*
import a.erubit.platform.course.CourseManager
import a.erubit.platform.course.lesson.CharacterLesson
import a.erubit.platform.course.lesson.Lesson
import a.erubit.platform.course.lesson.SetLesson
import a.erubit.platform.course.lesson.WelcomeLesson

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.StaggeredGridLayoutManager

import u.U

import java.util.*


class ProgressFragment : Fragment() {
	private var mListener: OnLessonInteractionListener? = null

	override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
		return inflater.inflate(R.layout.fragment_progress, container, false)
	}

	override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
		super.onViewCreated(view, savedInstanceState)

		val listView: RecyclerView = view.findViewById(android.R.id.list)
		listView.itemAnimator = DefaultItemAnimator()
		listView.layoutManager = StaggeredGridLayoutManager(1, StaggeredGridLayoutManager.VERTICAL)

		val course = CourseManager.i().getCourse(arguments!!.getString("id"))
		(view.findViewById<View>(R.id.courseName) as TextView).text = course!!.name
		listView.adapter = ProgressListAdapter(course)
	}

	override fun onAttach(context: Context) {
		super.onAttach(context)

		if (context is OnLessonInteractionListener)
			mListener = context
	}


	private inner class ProgressListAdapter internal constructor(course: Course?) : RecyclerView.Adapter<RecyclerView.ViewHolder?>() {
		private val mList: MutableList<ListItem>

		private fun getKnowledgeText(knowledgeLevel: Int): String {
			return when {
				knowledgeLevel > 0 && knowledgeLevel < SetLesson.RANK_FAMILIAR -> getString(R.string.studying)
				knowledgeLevel >= SetLesson.RANK_FAMILIAR && knowledgeLevel < SetLesson.RANK_LEARNED -> getString(R.string.familiar)
				knowledgeLevel >= SetLesson.RANK_LEARNED && knowledgeLevel < SetLesson.RANK_LEARNED_WELL -> getString(R.string.learned)
				knowledgeLevel >= SetLesson.RANK_LEARNED_WELL -> getString(R.string.learned_well)
				else -> getString(R.string.unknown)
			}
		}

		override fun getItemViewType(position: Int): Int {
			return mList[position].type
		}

		override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
			val inflater = LayoutInflater.from(parent.context)

			return when (viewType) {
				HEADER -> {
					val view = inflater.inflate(R.layout.item_progress_header, parent, false)
					HeaderViewHolder(view)
				}
				CONTENT -> {
					val view = inflater.inflate(R.layout.item_progress_content, parent, false)
					ContentViewHolder(view)
				}
				else -> throw UnsupportedOperationException("ViewType value `$viewType` is not supported.")
			}
		}

		override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
			if (holder is HeaderViewHolder) {
				holder.mLesson = (mList[position] as HeaderItem).mLesson
				holder.textTitle.text = (mList[position] as HeaderItem).mName
			}

			if (holder is ContentViewHolder) {
				val item = mList[position] as ContentItem
				holder.textTitle.text = U.defurigana(item.mText)
				holder.textKnowledge.text = item.mKnowledge
			}
		}

		override fun getItemCount(): Int {
			return mList.size
		}


		private inner class HeaderViewHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView) {
			var mLesson: Lesson? = null
			val textTitle: TextView = itemView.findViewById(android.R.id.text1)
            private val btnPractice: Button = itemView.findViewById(android.R.id.button1)

            init {
                btnPractice.setOnClickListener { mListener!!.onLessonInteraction(mLesson, LessonsFragment.LessonInteractionAction.PRACTICE) }
			}
		}

		private inner class ContentViewHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView) {
			val textTitle: TextView = itemView.findViewById(android.R.id.text1)
            val textKnowledge: TextView = itemView.findViewById(android.R.id.text2)
        }


        internal abstract inner class ListItem {
			abstract val type: Int
		}


		private inner class HeaderItem internal constructor(val mLesson: Lesson) : ListItem() {
			val mName: String? = mLesson.name
			override val type: Int
				get() = Companion.HEADER

		}


		private inner class ContentItem internal constructor(val mText: String, val mKnowledge: String) : ListItem() {
			override val type: Int
				get() = Companion.CONTENT
		}


		init {
			mList = ArrayList(10)
            val ctx = context!!

			for (lesson in course!!.lessons!!) {
				mList.add(HeaderItem(lesson))

				if (lesson is WelcomeLesson) {
					mList.add(ContentItem(lesson.getProgress(ctx).getExplanation(ctx), ""))
				}

				if (lesson is CharacterLesson) {
                    // including SetLesson, VocabularyLesson, PhraseLesson
					for (item in lesson.mSet!!) {
						mList.add(ContentItem((item as CharacterLesson.Item).character, getKnowledgeText(item.knowledgeLevel)))
					}
				}
			}
		}
	}


	companion object {
		const val HEADER = 0
		const val CONTENT = 1
	}

}
